oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,393 @@
+use crate::traits::HistogramSupport;
+use num::traits::Float;
+use num::traits::ToPrimitive;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use std::cmp::min;
+use std::ops::Sub;
+use thiserror::Error;
+
+/// Errors related to constructing a `Quantile` instance or estimating the p-quantile.
+#[derive(Debug, Clone, Error, JsonSchema, Serialize, Deserialize)]
+#[serde(tag = "type", content = "content", rename_all = "snake_case")]
+pub enum QuantileError {
+    /// The p value must be in the range [0, 1].
+    #[error("The p value must be in the range [0, 1].")]
+    InvalidPValue,
+    /// Quantile estimation is not possible without samples.
+    #[error("Quantile estimation is not possible without samples.")]
+    EmptyQuantile,
+    /// A non-finite was encountered, either as a bin edge or a sample.
+    #[error("Samples must be finite values, found: {0:?}")]
+    NonFiniteValue(String),
+}
+
+/// Structure for estimating the p-quantile of a population.
+///
+/// This is based on the P² algorithm for estimating quantiles using
+/// constant space.
+///
+/// The algorithm consists of maintaining five markers: the
+/// minimum, the p/2-, p-, and (1 + p)/2 quantiles, and the maximum.
+///
+/// Read <https://www.cs.wustl.edu/~jain/papers/ftp/psqr.pdf> for more.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
+pub struct Quantile {
+    /// The heights of the markers.
+    #[serde(rename = "marker_heights")]
+    pub marker_hghts: [f64; 5],
+    /// The positions of the markers.
+    #[serde(rename = "marker_positions")]
+    pub marker_pos: [i64; 5],
+    /// The desired marker positions.
+    #[serde(rename = "desired_marker_positions")]
+    pub desired_marker_pos: [f64; 5],
+    /// The increment for the desired marker positions.
+    #[serde(rename = "desired_marker_increments")]
+    pub desired_marker_incrs: [f64; 5],
+}
+
+impl Quantile {
+    /// Create a new `Quantile` instance.
+    pub fn new(p: u64) -> Result<Self, QuantileError> {
+        let p = p as f64 / 100.;
+        if p < 0. || p > 1. {
+            return Err(QuantileError::InvalidPValue);
+        }
+
+        Ok(Self {
+            marker_hghts: [0.; 5],
+            marker_pos: [1, 2, 3, 4, 0],
```

Comment:
Yeah, I'll double-check on this bit.

Reply:
Ok, reworked to code to not do this.

Reply:
Revisited this in a few implementations and the paper. It actually makes complete sense as the 5 marker position mirrors the observation after it because useful once filled.

Reply:
This position gets updated accordingly after we've filled it. Will document.